### PR TITLE
Fix on pagination

### PR DESCRIPTION
**Pull request by LenaigKaliou**

### CODE REVIEW

File: Lacework/lacework_module/lacework_connector.py [hunk 1]
```diff
@@ -38,32 +39,31 @@ class LaceworkEventsTrigger(Connector):
     def __init__(self, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
         self.from_date = ""
-        self.context = PersistentJSON("context.json", self._data_path)
+        self.context = PersistentJSON("context.json", "./")
```
Comment (TOUFIKIzakarya) [inline Lacework/lacework_module/lacework_connector.py:42]:
```suggestion
        self.context = PersistentJSON("context.json", self._data_path)
```

**Review by TOUFIKIzakarya — DISMISSED** [review]
Thanks lenaïg, this's a little remarks and also I suggest having different logs for the two methods `get_next_events` and `get_response_by_timestamp`. Like `Request on Lacework Central API to fetch next page events failed with` for `get_next_events`

**Review by TOUFIKIzakarya — APPROVED** [review]
LGFM

File: Lacework/lacework_module/lacework_connector.py [hunk 2]
```diff
@@ -36,17 +38,32 @@ class LaceworkEventsTrigger(Connector):
...
+            return most_recent_date_seen
 
-    @cursor.setter
-    def cursor(self, cursor: str) -> None:
+    @most_recent_date_seen.setter
+    def most_recent_date_seen(self, recent_date: str) -> None:
+        self.from_date = recent_date
         with self.context as cache:
-            cache["cursor"] = cursor
+            cache["timestamp_cursor"] = self.from_date
```
Comment (squioc) [inline Lacework/lacework_module/lacework_connector.py:66]: The getter gets its value from the key `most_recent_date_seen` in the cache
```suggestion
            cache["most_recent_date_seen"] = self.from_date
```

File: Lacework/lacework_module/lacework_connector.py [hunk 3]
```diff
@@ -121,55 +138,58 @@ def get_next_events(self, cursor: str | None) -> Any | None:
...
+            grouped_data = []
+            while next_page_url:
+                response_next_page = self.get_next_events(next_page_url)
+                if response_next_page:
+                    next_page_url = (
+                        response_next_page.get("content", {}).get("paging", {}).get("urls", {}).get("nextPage")
+                    )
+                    next_page_items = response_next_page.get("items", [])
+                    next_page_items = [orjson.dumps(message).decode("utf-8") for message in next_page_items]
+                    grouped_data.extend(next_page_items)
```
Comment (squioc) [inline Lacework/lacework_module/lacework_connector.py:179]: We have to compute the lag for each batch
```suggestion
                    grouped_data.extend(next_page_items)
                     last_date = self._get_most_recent_timestamp_from_items(next_page_items)
                    events_lag = int(time.time() - last_date)
                    EVENTS_LAG.labels(intake_key=self.configuration.intake_key).set(events_lag)
                    self.most_recent_date_seen = last_date
```

File: Lacework/lacework_module/lacework_connector.py [hunk 4]
```diff
@@ -121,55 +138,58 @@ def get_next_events(self, cursor: str | None) -> Any | None:
...
+                        OUTCOMING_EVENTS.labels(intake_key=self.configuration.intake_key).inc(len(grouped_data))
+                        self.push_events_to_intakes(events=grouped_data)
+                        grouped_data = []
+
+            if grouped_data != []:
+                self.log(message=f"Sending a batch of {len(grouped_data)} messages", level="info")
+                OUTCOMING_EVENTS.labels(intake_key=self.configuration.intake_key).inc(len(grouped_data))
+                self.push_events_to_intakes(events=grouped_data)
+
+            self.most_recent_date_seen = now
```
Comment (squioc) [inline Lacework/lacework_module/lacework_connector.py:192]: `most_recent_date_seen` must host the most recent date seen in the events. I suggest changing the way to value this property
```suggestion
```

File: Lacework/lacework_module/lacework_connector.py [hunk 5]
```diff
@@ -121,55 +138,58 @@ def get_next_events(self, cursor: str | None) -> Any | None:
...
 
-            items = batch.get("data", [])
-            if len(items) > 0:
-                most_recent_timestamp_seen = self._get_most_recent_timestamp_from_items(items)
-                events_lag = int(time.time() - most_recent_timestamp_seen)
-                EVENTS_LAG.labels(intake_key=self.configuration.intake_key).set(events_lag)
-                INCOMING_EVENTS.labels(intake_key=self.configuration.intake_key).inc(len(items))
+            last_date = self._get_most_recent_timestamp_from_items(items)
+            events_lag = int(time.time() - last_date)
+            EVENTS_LAG.labels(intake_key=self.configuration.intake_key).set(events_lag)
```
Comment (squioc) [inline Lacework/lacework_module/lacework_connector.py:163]: I suggest saving the most recent date seen here for the first batch
```suggestion
            EVENTS_LAG.labels(intake_key=self.configuration.intake_key).set(events_lag)
            self.most_recent_date_seen = last_date
```

File: Lacework/lacework_module/lacework_connector.py [hunk 6]
```diff
@@ -36,17 +38,32 @@ class LaceworkEventsTrigger(Connector):
...
+            if most_recent_date_seen < one_week_ago:
+                most_recent_date_seen = one_week_ago
+
+            return most_recent_date_seen
 
-    @cursor.setter
-    def cursor(self, cursor: str) -> None:
+    @most_recent_date_seen.setter
+    def most_recent_date_seen(self, recent_date: str) -> None:
+        self.from_date = recent_date
```
Comment (squioc) [inline Lacework/lacework_module/lacework_connector.py:64]: `from_date` must be the greatest date seen. Add a comparison to only save increasing dates
```suggestion
        # save the greatest date
        if self.from_date < recent_date:
                self.from_date = recent_date
```

File: Lacework/lacework_module/lacework_connector.py [hunk 7]
```diff
@@ -121,55 +138,58 @@ def get_next_events(self, cursor: str | None) -> Any | None:
...
-        cursor = self.cursor
-
-        messages = []
-        while has_more_messages and self.running:
-            has_more_messages = False
-
-            batch = self.get_next_events(cursor)
-            if batch is None:
-                break
+        now = datetime.now(timezone.utc)
```
Comment (squioc) [inline Lacework/lacework_module/lacework_connector.py:154]: Not sure that it's still used according to the suggestions!?
```suggestion
```

**Review by squioc — CHANGES_REQUESTED** [review]
Thanks for the fixes.
Still some changes to apply

**Review by squioc — APPROVED** [review]
LGTM

**Review by squioc — CHANGES_REQUESTED** [review]
Some changes to apply.

File: Lacework/lacework_module/lacework_connector.py [hunk 8]
```diff
@@ -36,17 +37,34 @@ class LaceworkEventsTrigger(Connector):
...
+            if most_recent_date_seen.replace(tzinfo=timezone.utc) < one_week_ago:
+                most_recent_date_seen = one_week_ago
+
+            return most_recent_date_seen
+
+    @most_recent_date_seen.setter
+    def most_recent_date_seen(self, recent_date: datetime) -> None:
+        # save the greatest date
+        if self.from_date == "" or isoparse(self.from_date) < recent_date:
+            self.from_date = str(recent_date)
```
Comment (squioc) [inline Lacework/lacework_module/lacework_connector.py:65]: To represent a datetime as a string, it's preferable to use [`isoformat`](https://docs.python.org/3/library/datetime.html#datetime.datetime.isoformat):
```suggestion
            self.from_date = recent_date.isoformat()
```

**Review by squioc — DISMISSED** [review]
Still a little change to apply

**Review by squioc — APPROVED** [review]
LGTM

**Pull request merged**